homeassistant/components/rotel/media_player.py

```diff
@@ -0,0 +1,493 @@
+        config.get(CONF_NAME),
+        config.get(CONF_DEVICE),
+        config.get(CONF_SOURCE_ALIASES))
+
+    async def handle_hass_stop_event(event):
+        """Clean up when hass stops."""
+        await device.cleanup()
+
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        handle_hass_stop_event)
+    _LOGGER.debug("Registered device '%s' for HASS stop event", device.name)
+
+    await device.open_connection()
+
+    async_add_entities([device])
+    device.start_read_messages(hass)
+    setup_hass_services(hass)
+
+
+def setup_hass_services(hass):
+    """
+    Register services.
+
+    Note that this function is called for every entity but it
+    only needs to be called once for the platform.
+    It doesn't seem to do any harm but I'd like to tidy that up at some point.
+    """
+    _LOGGER.debug("Setting up hass services")
+
+    async def async_handle_send_command(entity, call):
+        command_name = call.data.get(ATTR_COMMAND_NAME)
+        if isinstance(entity, RotelMediaPlayer):
+            _LOGGER.debug(
+                "%s service sending command %s to entity %s",
+                SERVICE_SEND_COMMAND,
+                command_name,
+                entity.entity_id)
+            await entity.send_command(command_name)
+        else:
+            _LOGGER.debug(
+                "%s service not sending command %s to incompatible entity %s",
+                SERVICE_SEND_COMMAND,
+                command_name,
+                entity.entity_id)
+
+    async def async_handle_reconnect(entity, call):
+        # pylint: disable=unused-argument
+        if isinstance(entity, RotelMediaPlayer):
+            _LOGGER.debug("%s service reconnecting entity %s",
+                          SERVICE_RECONNECT, entity.entity_id)
+            await entity.reconnect()
+        else:
+            _LOGGER.debug("%s service not reconnecting incompatible entity %s",
+                          SERVICE_RECONNECT, entity.entity_id)
+
+    component = hass.data[DOMAIN]
+    component.async_register_entity_service(
+        SERVICE_SEND_COMMAND, SERVICE_SEND_COMMAND_SCHEMA,
+        async_handle_send_command)
+    component.async_register_entity_service(
+        SERVICE_RECONNECT, SERVICE_RECONNECT_SCHEMA,
+        async_handle_reconnect)
+
+
+class RotelMediaPlayer(MediaPlayerDevice):
+    """Representation of a Rotel media player."""
+
+    # pylint: disable=abstract-method
+    # pylint: disable=too-many-public-methods
+    # pylint: disable=too-many-instance-attributes
+
+    def __init__(self, name, device, source_aliases):
+        """Initialize the device."""
+        self._name = name
+        self._device = device
+        self._conn = None  # Make sure that you call open_connection...
+        self._read_messages_task = None  # ... and start_read_messages
+        self._state = STATE_OFF
+        self._source_name = None
+        self._volume = None
+        self.set_source_lists(source_aliases)
```

Comment:
This causes side-effects, which we do not want in `__init__`, which can be solved by removing the debug calls in the method -- or call it outside of `__init__`.